From 3d843f3f8e4678f2f92a20e21cf34bea56f4be76 Mon Sep 17 00:00:00 2001
From: Alex Kanter <alex.kanter@devalore.com>
Date: Sun, 21 Feb 2021 14:55:08 +0200
Subject: [PATCH] add an option to set secondary channel

Ported commit from UGW (WLANRTSYS-30243).
Needed for cac termination feature  to pass 4.16.1 easymesh R2 test.

Signed-off-by: Alex Kanter <alex.kanter@devalore.com>
---
 hostapd/config_file.c | 9 +++++++--
 src/ap/ap_config.h    | 1 +
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/hostapd/config_file.c b/hostapd/config_file.c
index be0b2c4..4c6e541 100644
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -1400,11 +1400,13 @@ static int hostapd_config_ht_capab(struct hostapd_config *conf,
 		conf->ht_capab |= HT_CAP_INFO_LDPC_CODING_CAP;
 	if (os_strstr(capab, "[HT40-]")) {
 		conf->ht_capab |= HT_CAP_INFO_SUPP_CHANNEL_WIDTH_SET;
-		conf->secondary_channel = -1;
+		if (!conf->secondary_channel_set)
+			conf->secondary_channel = -1;
 	}
 	if (os_strstr(capab, "[HT40+]")) {
 		conf->ht_capab |= HT_CAP_INFO_SUPP_CHANNEL_WIDTH_SET;
-		conf->secondary_channel = 1;
+		if (!conf->secondary_channel_set)
+			conf->secondary_channel = 1;
 	}
 	if (os_strstr(capab, "[HT40+]") && os_strstr(capab, "[HT40-]")) {
 		conf->ht_capab |= HT_CAP_INFO_SUPP_CHANNEL_WIDTH_SET;
@@ -4493,6 +4495,9 @@ int hostapd_config_fill(struct hostapd_config *conf,
 				   line);
 			return 1;
 		}
+	} else if (os_strcmp(buf, "secondary_channel") == 0) {
+		conf->secondary_channel = atoi(pos);
+		conf->secondary_channel_set = 1;
 	} else if (os_strcmp(buf, "ht_tx_bf_capab_from_hw") == 0) {
 		conf->ht_tx_bf_capab_from_hw = atoi(pos);
 	} else if (os_strcmp(buf, "ht_tx_bf_capab") == 0) {
diff --git a/src/ap/ap_config.h b/src/ap/ap_config.h
index c1daae1..60bfd30 100644
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1026,6 +1026,7 @@ struct hostapd_config {
 	int ieee80211n;
 	int ieee80211n_acax_compat;
 	int secondary_channel;
+	int secondary_channel_set = 0;
 	int no_pri_sec_switch;
 	int ht_rifs;
 	int require_ht;
-- 
2.17.1

